File: 92.cpp

#include <bits/stdc++.h>
#include <errno.h>
#include <fcntl.h>
#include <sys/types.h>
#include <unistd.h>

using namespace std;

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  ListNode *reverseBetween(ListNode *head, int left, int right) {
    int pos = 1;
    ListNode *dummy = new ListNode;
    dummy->next = head;
    ListNode *pre = dummy;
    ListNode *cur = head;
    while (cur && pos < left) {
      pre = pre->next;
      cur = cur->next;
      pos++;
    }
    ListNode *tailNode = cur;
    while (cur && pos <= right) {
      ListNode *next = cur->next;
      cur->next = pre->next;
      pre->next = cur;
      cur = next;
      tailNode->next = next;
      pos++;
    }
    head = dummy->next;
    delete dummy;
    return head;
  }
};

int main(int argc, char const *argv[]) {
  Solution s;

  return 0;
}

File: 92.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseBetween(head *ListNode, left int, right int) *ListNode {
	var index int
	dummy := &ListNode{Next: head}
	var cur *ListNode = head
	var prev *ListNode = dummy
	left -= 1
	right -= 1
	for index < left {
		temp := cur
		cur = cur.Next
		prev = temp
		index++
	}
	beginPrev := prev
	begin := cur
	for index < right {
		temp := cur
		tempNext := cur.Next
		cur.Next = prev
		prev = temp
		cur = tempNext
		index++
	}
	endNext := cur.Next
	end := cur
	end.Next = prev
	begin.Next = endNext
	beginPrev.Next = end
	return dummy.Next
}

func main() {
	nodes := []ListNode{
		ListNode{1, nil},
		ListNode{2, nil},
		ListNode{3, nil},
		ListNode{4, nil},
		ListNode{5, nil},
	}
	nodes[0].Next = &nodes[1]
	nodes[1].Next = &nodes[2]
	nodes[2].Next = &nodes[3]
	nodes[3].Next = &nodes[4]
	head := reverseBetween(&nodes[0], 2, 4)
	for head != nil {
		fmt.Print(head.Val)
		head = head.Next
	}
}
